test/jdk/javax/swing/JTable/JTableRightAlignmentTest.java

```diff
@@ -0,0 +1,217 @@
+import java.awt.Insets;
+import java.awt.Rectangle;
+import java.awt.Robot;
+import java.awt.image.BufferedImage;
+
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.JTable;
+import javax.swing.SwingUtilities;
+import javax.swing.table.AbstractTableModel;
+import javax.swing.UIManager;
+import javax.swing.UnsupportedLookAndFeelException;
+
+import javax.imageio.ImageIO;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+ * @test
+ * @key headful
+ * @bug 5108458
+ * @summary Test to check Right alignment of JTable data
+ * (Fix affects all L&F, test verifies only Metal L&F)
+ * @run main JTableRightAlignmentTest
+ */
+
+public class JTableRightAlignmentTest {
+    static JFrame frame;
+    static CustomTable table;
+    static String failureString;
+
+    public static void main(String[] args) throws Exception {
+        Robot robot = new Robot();
+        for (UIManager.LookAndFeelInfo laf : UIManager.getInstalledLookAndFeels()) {
+            System.out.println("Testing LAF : " + laf.getClassName());
+            robot.waitForIdle();
+            robot.delay(100);
+            SwingUtilities.invokeAndWait(() -> setLookAndFeel(laf));
+            try {
+                SwingUtilities.invokeAndWait(() -> {
+                    frame = new JFrame("Test JTable");
+                    JPanel panel = new JPanel(new GridBagLayout());
+                    frame.setContentPane(panel);
+                    table = new CustomTable();
+                    panel.add(new JScrollPane(table.table),
+                            new GridBagConstraints(0, 0, -1, -1, 1.0, 1.0,
+                                    GridBagConstraints.PAGE_START, GridBagConstraints.BOTH,
+                                    new Insets(2, 2, 2, 2), 0, 0));
+                    frame.pack();
+                    frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
+                    frame.applyComponentOrientation(ComponentOrientation.RIGHT_TO_LEFT);
+                    frame.setVisible(true);
+                });
+
+                robot.waitForIdle();
+                robot.delay(1000);
+                SwingUtilities.invokeAndWait(() -> {
+                    int allColumnWidths = 0;
+                    for (int i = 0; i < table.table.getColumnCount(); i++) {
+                        allColumnWidths += table.table.getTableHeader().getColumnModel()
+                                .getColumn(i)
+                                .getWidth();
+                    }
+                    BufferedImage bufferedImage = robot.createScreenCapture(
+                            new Rectangle(table.table.getLocationOnScreen().x,
+                                    table.table.getLocationOnScreen().y,
+                                    table.table.getWidth() - allColumnWidths,
+                                    table.table.getHeight()));
```

Comment:
This is the reason why I dislike the model design: `table.table` doesn't look good and is somewhat confusing. Moreover, `CustomTable` is not really _a table_ which makes the design even more confusing: a table contains a table.

I understand over-engineering a test is a waste of time, yet it feels too wrong this way. A clean test code will ease the work of another engineer who will need to deal with updating the test if anything changes.

Reply:
Will work on it then.